src/unix/futex.c

```diff
@@ -294,3 +294,106 @@ init_futices(process p)
         halt("failed to allocate futex table\n");
 
 }
+
+/* robust mutex handling */
+
+#define FUTEX_OWNER_DIED	0x40000000
+#define FUTEX_KEY_ADDR(x, o)    ((u32 *)((u8 *)(x) + (o)))
+
+typedef struct robust_list {
+    struct robust_list *next;
+} *robust_list;
+
+typedef struct robust_list_head {
+    robust_list list;
+    long futex_offset;
+    void *list_op_pending;
+} *robust_list_head;
+
+static boolean valid_user_address(process p, void *a, int len)
+{
+    u64 pg = u64_from_pointer(a) & ~PAGEMASK;
+    u64 end = (pg + len) & ~PAGEMASK;
+
+    for (; pg <= end; pg += PAGESIZE) {
+        if (vmap_from_vaddr(p, pg) == INVALID_ADDRESS)
+                return false;
+    }
+    return true;
+}
+
+void wake_robust_list(process p, void *head)
+{
+    struct futex * f;
+    struct robust_list_head *h = head;
+    struct robust_list *l;
+    u32 *uaddr;
+
+    /* must be very careful accessing the head as well as the list */
+    if (!valid_user_address(p, h, sizeof *h))
+        return;
+
+    /* XXX could keep a list of futexes and wake them at the end
+     * to let threads acquire multiple locks without blocking */
+    if (h->list_op_pending) {
+        uaddr = FUTEX_KEY_ADDR(h->list_op_pending, h->futex_offset);
+        if (valid_user_address(p, uaddr, sizeof *uaddr)) {
+            *uaddr |= FUTEX_OWNER_DIED;
+            if ((f = table_find(p->futices, uaddr)))
+                futex_wake_many(f, 1);
```

Comment:
Shouldn't all waiters (instead of just one) be woken up? I think we should use INT_MAX (which is the value used by FUTEX_WAKE to indicate that all waiters should be woken up) instead of 1.
Also, we could use futex_wake_many_by_uaddr(), which does the table lookup.

Reply:
I went with one since that's how it's described in Ingo Molnar's 2006 email about kernel behavior for robust lists, and that's also what pthread_mutex_unlock uses for the futex wake syscall. I think it makes sense to just wake the next in queue blocked on the lock. I might be missing something though...what situation are you thinking of where waking all would be beneficial?
Thanks for the tip about futex_wake_many_by_uaddr!

Reply:
I was under the assumption that a futex held by a dead thread wasn't usable anymore, which is of course not the case. Now I think you are right, it makes more sense to wake up just one blocked thread